src/test/java/at/ac/tuwien/kr/alpha/solver/SolverStatisticsTests.java

```diff
@@ -31,32 +31,75 @@
 import java.util.Set;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assume.assumeTrue;
 
 public class SolverStatisticsTests extends AbstractSolverTests {
 
 	@Test
 	public void checkStatsStringZeroChoices() {
 		Solver solver = getInstance("a.");
+		assumeTrue(solver instanceof SolverMaintainingStatistics);
 		collectAnswerSetsAndCheckStats(solver, 1, 0, 0, 0, 0, 0, 0, 0);
 	}
 
 	@Test
 	public void checkStatsStringOneChoice() {
 		Solver solver = getInstance("a :- not b. b :- not a.");
+		assumeTrue(solver instanceof SolverMaintainingStatistics);
 		collectAnswerSetsAndCheckStats(solver, 2, 1, 1, 1, 1, 0, 0, 0);
 	}
 
+	@Test
+	public void checkNoGoodCounterStatsByTypeZeroChoices() {
+		Solver solver = getInstance("a.");
+		assumeTrue(solver instanceof SolverMaintainingStatistics);
+		collectAnswerSetsAndCheckNoGoodCounterStatsByType(solver, 0, 0, 0, 0);
+	}
+
+	@Test
+	public void checkNoGoodCounterStatsByTypeOneChoice() {
+		Solver solver = getInstance("a :- not b. b :- not a.");
+		assumeTrue(solver instanceof SolverMaintainingStatistics);
+		collectAnswerSetsAndCheckNoGoodCounterStatsByType(solver, 7, 2, 0, 4);
```

Comment:
The idea of these tests is to check if statistics are correct for the current implementation of `NoGoodGenerator`. If this implementation changes, these tests might have to change as well. I expect that this will happen very rarely.

Reply:
I expect that to happen soon, so I would like to avoid those tests failing then. Isn't there another way to get a sure count on the number of actual nogoods?

Reply:
As far as I understand, you want to compare the actual numbers of nogoods to the statistics produced by `NoGoodCounter`. This would be a different kind of test, because currently the expected numbers of nogoods are compared to those statistics. In my opinion, the current test is valuable because Alpha produces nogoods in a well-documented and stable way.

To conduct the kind of test you propose, it would be necessary to obtain the `NoGoodStore` instance from the solver (for which currently no non-private method exists), and it would be necessary to obtain the nogoods from the store and count them by an independent procedure. Obtaining nogoods from the store is also not supported by any non-private method currently, and is complicated by the fact that the store does not even contain a single collection of nogoods (e.g., unary nogoods are not stored at all, binary nogoods are stored with their respective watches, ...).

Reply:
Well, I am not sure what these tests should actually check. From the name of the class I was expecting it to be unit tests regarding the reported statistics. What it currently does partially, is verifying behaviour of the `NoGoodGenerator`, which should be in another class. Further changes, like the stratified evaluation, will also influence the number of nogoods reported to the solver. I am fine with checking the statistics-handling, but I think that should not depend as much on current internal behaviour as it does right now.

I agree that actually verifying the statistics involves a rather complex setup (and I am not fully sure if the test coverage is worth the effort). I would expect that a tailored `Grounder` could do the job (like the `ChoiceGrounder` or `DummyGrounder`) of guaranteeing that certain types of nogoods are handed to the solver, and it might be sane to expect certain numbers in the statistics based on the nogoods obtained from the grounder.

Reply:
Good idea. I changed the tests to just use the existing `DummyGrounder` in 6df645a8bc8ed1694d3f1124efddbeb4ed3ff538 to keep the effort low.